tapchannel/aux_traffic_shaper.go

```diff
@@ -0,0 +1,242 @@
+// routing.TlvTrafficShaper interface.
+var _ routing.TlvTrafficShaper = (*AuxTrafficShaper)(nil)
+
+// HandleTraffic is called in order to check if the channel identified by the
+// provided channel ID may have external mechanisms that would allow it to carry
+// out the payment.
+func (s *AuxTrafficShaper) HandleTraffic(_ lnwire.ShortChannelID,
+	fundingBlob lfn.Option[tlv.Blob]) (bool, error) {
+
+	// If there is no auxiliary blob in the channel, it's not a custom
+	// channel, and we don't need to handle it.
+	if fundingBlob.IsNone() {
+		return false, nil
+	}
+
+	// If we can successfully decode the channel blob as a channel capacity
+	// information, we know that this is a custom channel.
+	err := lfn.MapOptionZ(fundingBlob, func(blob tlv.Blob) error {
+		_, err := cmsg.DecodeOpenChannel(blob)
+		return err
+	})
+	if err != nil {
+		return false, nil
+	}
+
+	// No error, so this is a custom channel, we'll want to decide.
+	return true, nil
+}
+
+// PaymentBandwidth returns the available bandwidth for a custom channel decided
+// by the given channel aux blob and HTLC blob. A return value of 0 means there
+// is no bandwidth available. To find out if a channel is a custom channel that
+// should be handled by the traffic shaper, the HandleTraffic method should be
+// called first.
+func (s *AuxTrafficShaper) PaymentBandwidth(htlcBlob,
+	commitmentBlob lfn.Option[tlv.Blob]) (lnwire.MilliSatoshi, error) {
+
+	// This method shouldn't be called if we don't have a commitment blob
+	// available (i.e., the channel is not a custom channel).
+	if commitmentBlob.IsNone() || htlcBlob.IsNone() {
+		return 0, fmt.Errorf("no commitment or TLV blob available " +
+			"for custom channel bandwidth estimation")
+	}
+
+	commitmentBytes := commitmentBlob.UnsafeFromSome()
+	htlcBytes := htlcBlob.UnsafeFromSome()
+
+	commitment, err := cmsg.DecodeCommitment(commitmentBytes)
+	if err != nil {
+		return 0, fmt.Errorf("error decoding commitment blob: %w", err)
+	}
+
+	htlc, err := rfqmsg.DecodeHtlc(htlcBytes)
+	if err != nil {
+		return 0, fmt.Errorf("error decoding HTLC blob: %w", err)
+	}
+
+	localBalance := cmsg.OutputSum(commitment.LocalOutputs())
+
+	// There either already is an amount set in the HTLC (which would
+	// indicate it to be a keysend payment), in which case we don't need an
+	// RFQ ID as we can just compare the local balance and the required HTLC
+	// amount. If there is no amount set, we need to look up the RFQ ID in
+	// the HTLC blob and use the accepted quote to determine the amount.
+	htlcAssetAmount := htlc.Amounts.Val.Sum()
+	if htlcAssetAmount != 0 && htlcAssetAmount <= localBalance {
+		// We signal "infinite" bandwidth by returning a very high
+		// value (number of Satoshis ever in existence).
+		return lnwire.NewMSatFromSatoshis(btcutil.MaxSatoshi), nil
+	}
+
+	// If the HTLC doesn't have an RFQ ID, we can't use any channel.
+	if htlc.RfqID.ValOpt().IsNone() {
```

Comment:
shouldn't this check be done earlier?

Reply:
No. For a direct peer keysend, we have an asset amount encoded in the HTLC custom records and can use that directly. So if we did this check earlier, we would return an error for a direct peer keysend that doesn't have an RFQ ID (related to your question below, see answer there).